working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+omitted in the case where the immediaty syntactic context implies that this
+modifier _must_ be present. For example `const [const C()]` can be written
+as `const [C()]`. In the examples above, the parameter-and-variable
+declarations `final int x` and `final int y` are written as `int x` and
+`int y`, and this is allowed because it would be a compile-time error to
+omit `final` in an `inline` class, and in a class with a constant
+constructor. In other words, when we see `inline` on the class or `const`
+on the class name, we know that `final` is implied on all instance
+variables.
+
+A proposal which was mentioned during the discussions about primary
+constructors was that the keyword `final` could be used in order to specify
+that all instance variables introduced by the primary constructor are
+`final` (but the constructor wouldn't be constant, and hence there's more
+freedom in the declaration of the rest of the class). However, that
+proposal is not included here, because it may be a source of confusion that
+`final` may also occur as a modifier on the class itself, and also because
+the resulting class header does not contain syntax which is already similar
+to a non-primary constructor declaration.
+`class final Point(int x, int y);` cannot use the similarity to a
+non-primary constructor declaration to justify the keyword `final`.
+
+```dart
+class Point {
+  final int x;
+  final int y;
+  Point(this.x, this.y);
+}
+
+class final Point(int x, int y); // Not supported!
+```
+
+Optional parameters can be declared as usual in a primary constructor, with
+default values that must be constant as usual:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, [this.y = 0]);
+}
+
+class Point(int x, [int y = 0]);
+```
+
+Similarly for named parameters, required or not:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, {required this.y});
+}
+
+class Point(int x, {required int y});
+```
+
+The class header can have additional elements, just like class headers
+where there is no primary constructor:
+
+```dart
+class D<TypeVariable extends Bound> extends A with M implements B, C {
+  final int x;
+  final int y;
+  const D.named(this.x, [this.y = 0]);
+}
+
+class const D.named<TypeVariable extends Bound>(int x, [int y = 0])
+    extends A with M implements B, C;
+```
+
+## Specification
+
+### Syntax
+
+The grammar is modified as follows. Note that the changes include
+support for inline classes, because they're intended to use primary
+constructors as well.
+
+```
+ <topLevelDefinition> ::=
+     <classDeclaration>
+   | <inlineClassDeclaration> // New alternative.
+   | ...;
+
+<classDeclaration> ::= // First alternative modified.
+     (<classModifiers> | <mixinClassModifiers>)
+     'class' <classNamePart> <superclass>? <interfaces>? <classBody>
+   | ...;
+
+<classNamePart> ::= // New rule.
+     'const'? <constructorName> <typeParameters>? <formalParameterList>
```

Comment:
I think the position of type parameters should be `C<T>.foo(args)`, not `C.foo<T>(args)`.
That's consistent with how the constructor is called.

So 
```ebnf
    'const'? <identifier><typeParameters>?('.' (<identifier>|'new'))? <formalParameterList>
```

Reply:
Maybe. I tried that, but it doesn't actually read very well:

```dart
// Follow the constructor invocations.
class C<X extends List<Y>, ... /*lots of stuff*/>.name(Y y) {...}

// Allow `C.name` to occur together.
class C.name<X extends List<Y>, ... /*lots of stuff*/>(Y y) {...}
```

It is certainly true that the first choice is more consistent with invocations, but I suspect that it matters more that `C.name` should be searchable, as it would be when the declaration is a body constructor.

So I did not change that for now, that sounds like a big dog fight in an issue. ;-)

Reply:
I'm a cat person, so I'll aim for a cat fight. 😸